src/emulator/kernel/src/thread.cpp

```diff
@@ -16,16 +16,45 @@
 // 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 
 #include <kernel/thread_functions.h>
-
 #include <kernel/thread_state.h>
+#include <psp2/types.h>
+#include <psp2/kernel/error.h>
+
+#include <kernel/state.h>
 
 #include <cpu/functions.h>
 #include <util/resource.h>
+#include <util/lock_and_find.h>
+#include <util/find.h>
+
+#include <SDL_thread.h>
 
 #include <cassert>
 #include <cstring>
 
-ThreadStatePtr init_thread(Ptr<const void> entry_point, size_t stack_size, bool log_code, MemState &mem, CallImport call_import) {
+struct ThreadParams {
+    KernelState *kernel = nullptr;
+    SceUID thid = SCE_KERNEL_ERROR_ILLEGAL_THREAD_ID;
+    SceSize arglen = 0;
+    Ptr<void> argp;
+    std::shared_ptr<SDL_semaphore> host_may_destroy_params = std::shared_ptr<SDL_semaphore>(SDL_CreateSemaphore(0), SDL_DestroySemaphore);
```

Comment:
You seem to only use `.get` so this can probably be a `unique_ptr`. Please always prefer that, or `weak_ptr` over `shared_ptr`.

Reply:
My fault, I overuse shared_ptr because you don't need to specify the deleter type for incomplete types, as you do for unique_ptr (I think.)

Reply:
Yeah `unique_ptr` uses static deleters so the deleter type needs to be complete.

Reply:
Someone could explain me what are all those ptr 😢

Reply:
If I change it to a unique_ptr then thread_function will complain of call to implicitly-deleted copy constructor

Reply:
I see PetMac already answered :p

Reply:
I have problems when using unique_ptr. Like there's no copy constructor. any idea?